### ISSUE TITLE
[L0] Event dependency bug

### ISSUE THREAD

**Issue by AidanBeltonS** [intel/llvm#8132]
**Description**
I have ran into an issue with level zero's event behaviour.
I have found that an event which depends on another event is completed before it should.
It is unclear if this is an issue with the event info or the kernel is not waiting on the prior event.
There is a reproducer below. I have also included some hardware information. Let me know if you require any additional details.
The problem only exists with the level zero backend and does not occur with OpenCL.

**Reproducer**
The code below creates two tasks. The first copies memory from the host to the device, the second executes a basic kernel.
The code uses USM. We create a dependency between the fist task and the second.

The events status are checked to make sure the second task waits on the first before executing.
To do this the depender status is checked in a while loop until it is either running or complete.
Then once the first task is running or finished the second task is checked to see if it has completed yet.

If it does an error is created. We would expect that each backend would not throw the error.

```
#include <CL/sycl.hpp>

void check_dependency(cl::sycl::event dependee_e, cl::sycl::event depender_e) {
  cl::sycl::info::event_command_status depender_status =
      depender_e.get_info<cl::sycl::info::event::command_execution_status>();

  while (depender_status != cl::sycl::info::event_command_status::running &&
         depender_status != cl::sycl::info::event_command_status::complete) {
    depender_status =
        depender_e.get_info<cl::sycl::info::event::command_execution_status>();
  }
  if (dependee_e.get_info<cl::sycl::info::event::command_execution_status>() !=
      cl::sycl::info::event_command_status::complete) {
    throw std::runtime_error("Oh no!");
  }
}

int main(int argc, char const *argv[]) {
  sycl::queue q{};
  constexpr int n = 10'000'000;
  float *d_A = sycl::malloc_device<float>(n, q);
  float *h_A = sycl::malloc_host<float>(n, q);

  constexpr int n_repeat = 1000;

  for (int i = 0; i < n_repeat; ++i) {
    auto ev1 = q.memcpy(d_A, h_A, n * sizeof(float));
    auto ev2 = q.submit([&](sycl::handler &cgh) {
      cgh.depends_on({ev1});
      cgh.parallel_for(sycl::nd_range<1>(32, 32), [=](sycl::nd_item<1> item) {
        int x = item.get_global_linear_id();
        d_A[x] = x;
      });
    });
    check_dependency(ev1, ev2);
    q.wait_and_throw();
  }

  std::cout << "Success\n";

  sycl::free(d_A, q);
  sycl::free(h_A, q);
  return 0;
}
```

Results:
Level zero GPU
```
$ SYCL_DEVICE_FILTER=level_zero:gpu ./a.out
terminate called after throwing an instance of 'std::runtime_error'
  what():  Oh no!
Aborted (core dumped)
```

OpenCL GPU
```
$ SYCL_DEVICE_FILTER=opencl:gpu ./a.out
Success
```

OpenCL CPU
```
$ SYCL_DEVICE_FILTER=opencl:cpu ./a.out
Success
```

**Environment:**

- OS: 
```
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.5 LTS
Release:	20.04
Codename:	focal
```
- DPC++ version: 
 ```
 clang version 16.0.0 (https://github.com/AidanBeltonS/llvm/ 1fda2a37568b67ad929c703e0078dbe645e002cb)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/aidan/llvm/build/bin
 ```
- sycl-ls output:
[opencl:cpu:0] Intel(R) OpenCL, Intel(R) Core(TM) i7-10610U CPU @ 1.80GHz 3.0 [2022.13.3.0.16_160000]
[opencl:cpu:1] Intel(R) OpenCL, Intel(R) Core(TM) i7-10610U CPU @ 1.80GHz 3.0 [2022.13.3.0.16_160000]
[opencl:gpu:2] Intel(R) OpenCL HD Graphics, Intel(R) UHD Graphics [0x9b41] 3.0 [21.36.20889]
[ext_oneapi_level_zero:gpu:0] Intel(R) Level-Zero, Intel(R) UHD Graphics [0x9b41] 1.1 [1.2.20889]

**Comment by joeatodd** [comment]
We've also been able to reproduce this error on the CUDA backend, for what it's worth.

**Comment by abagusetty** [comment]
@AidanBeltonS This is an interesting one. The goal is that `e2` shouldn't execute until `e1` is complete. 

From you code snippet above, this logic seems to check for `depender_status` to be `event_command_status::submitted` state.
```c++
  while (depender_status != cl::sycl::info::event_command_status::running &&
         depender_status != cl::sycl::info::event_command_status::complete) {
    depender_status =
        depender_e.get_info<cl::sycl::info::event::command_execution_status>();
  }
```

It looks like `e2` is `submitted` and `e1` is `!complete`. This seems to be a valid state since L0 enqueues tasks in batches. The error state would be `e2` in `running` && `e1` in `!complete`. 

I think it might be okay to be in `submitted` state for `e2` while `e1` is still `running` or `submitted`.  It would be great if the spec clarifies a note.

It would be great if any one can confirm this.

```
info::event_command_status::submitted --> Indicates that the command has been submitted to the SYCL queue but has not yet started running on the device.
```

**Comment by AidanBeltonS** [comment]
So to clarify the behaviour. We aim to assert that `e1` is complete before `e2` starts running.
The goal behind this code is to allow us to write a unit test to check if our math functions are correct depending on a user's events.

The while loop checks to see if `e2` is either `running`/`completed`.
It breaks once it is in either of these states.

The code then checks if `e1` is `completed` and throws an error if it is not.

The expected behaviour is that `e1` is `completed` before `e2` is `running`/`completed`. This is because we have explicitly created a dependency between the two events.

**Comment by abagusetty** [comment]
@AidanBeltonS Apologies, I misunderstood the logic. I was able to test this issue on PVC with latest compute runtime (L0)and also on A100 and AMD devices. It fails there too as well. Sorry for the confusion. Could you please add CUDA and HIP tags to the title as well. Thanks!

**Comment by npmiller** [comment]
I'm also seeing this issue with the CUDA plugin, and I do think we should investigate.

But as a side note I don't think this snippet is technically valid SYCL, because I don't believe there is any guarantee that a submitted task will start executing or complete before some sort of `.wait()` is called, this was clarified in the SYCL spec recently, see:
* https://github.com/KhronosGroup/SYCL-Docs/pull/300/files#diff-3211f50334dd66e4638c43c11bb1b9dc16f0c26a8b7e5af104ea4e9f7b1744ecR1123

So even if in practice it's working with the current targets, there's no guarantee that `check_dependency` won't be an infinite loop.

**Comment by abagusetty** [comment]
Upon revisiting this section that is relevant to the topic: https://github.com/Pennycook/SYCL-Docs/blob/386749ce9ce4b372d7e574647c78955a1dd01d10/adoc/chapters/architecture.adoc#sycl-application-execution-model

I wasn't able to find a reference that enforces that a `synchronization` routine must be called to check on the status of the captured event. May be I missed. :) (totally possible) 
All I was able to find is:
``
When a [[command-group]] is submitted to a SYCL [[queue]], the requirements of the kernel execution are captured. The implementation can start executing a kernel as soon as its requirements have been satisfied.
``

Pinging @Pennycook @keryell if they can chime in reg. When the returned event can be quered for the status ?

**Comment by npmiller** [comment]
You're right, I don't think there's any issues with querying the event status here, what I'm concerned about is the `while` loop waiting for the event status to change, because I believe until a synchronization routine is called there's no guarantee that the commands will start executing, and so there's no guarantee the event status will change, so the `while` loop checking the event may be infinite on some platforms.

**Comment by abagusetty** [comment]
Thanks, I see it more clearly. I think the core question is: "whether a synchronization method is required for the commands to guarantee the start of execution and the events flagged as `running`". Sorry, should have been more clear. I was mostly trying to find for the mention of a `sync` requirement to guarantee for the command-group to start executing and its returned events to be meaningful. 

Yes I agree that a while loop check can lead to a infinite-loop depending on the device/platform/backend (in general)

**Comment by joeatodd** [comment]
I've been investigating this issue in more detail and I *think* the issue is in the mapping of backend event status to SYCL `event_command_status`.

### CUDA

From [pi_cuda.hpp](https://github.com/intel/llvm/blob/751acf357ed192a2f9de5171ccdf4a69510cff37/sycl/plugins/cuda/pi_cuda.hpp#L659-L669):

```c++
  pi_int32 get_execution_status() const noexcept {

    if (!is_recorded()) {
      return PI_EVENT_SUBMITTED;
    }

    if (!is_completed()) {
      return PI_EVENT_RUNNING;
    }
    return PI_EVENT_COMPLETE;
  }
```

Basically, as soon as the CUDA event is recorded onto the relevant stream, the SYCL event status changes from `submitted` to `running`. However, the recording of the CUDA event on the stream does *not* represent the task actually starting on the device.

A nice self-contained example comes from [this blog](https://developer.nvidia.com/blog/how-implement-performance-metrics-cuda-cc/):
```c++
cudaEvent_t start, stop;
cudaEventCreate(&start);
cudaEventCreate(&stop);

cudaMemcpy(d_x, x, N*sizeof(float), cudaMemcpyHostToDevice);
cudaMemcpy(d_y, y, N*sizeof(float), cudaMemcpyHostToDevice);

cudaEventRecord(start);
saxpy<<<(N+255)/256, 256>>>(N, 2.0f, d_x, d_y);
cudaEventRecord(stop);

cudaMemcpy(y, d_y, N*sizeof(float), cudaMemcpyDeviceToHost);

cudaEventSynchronize(stop);
float milliseconds = 0;
cudaEventElapsedTime(&milliseconds, start, stop);
```

The fact that the event has been recorded `cudaEventRecord()` simply indicates that the task has been pushed to the CUDA stream, it doesn't say anything about whether the event has actually happened.

### L0

I'm less familiar with level-zero, but the code from [pi_level_zero.cpp](https://github.com/intel/llvm/blob/3e8f937058f4f39b4762f947b05dbda4a59c6d4f/sycl/plugins/level_zero/pi_level_zero.cpp#L4802-L4841) is even more explicit about the possible event statuses:

```c++
  case PI_EVENT_INFO_COMMAND_EXECUTION_STATUS: {
    // Check to see if the event's Queue has an open command list due to
    // batching. If so, go ahead and close and submit it, because it is
    // possible that this is trying to query some event's status that
    // is part of the batch.  This isn't strictly required, but it seems
    // like a reasonable thing to do.
    auto Queue = Event->Queue;
    if (Queue) {
      // Lock automatically releases when this goes out of scope.
      std::scoped_lock<pi_shared_mutex> lock(Queue->Mutex);
      const auto &OpenCommandList = Queue->eventOpenCommandList(Event);
      if (OpenCommandList != Queue->CommandListMap.end()) {
        if (auto Res = Queue->executeOpenCommandList(
                OpenCommandList->second.isCopy(Queue)))
          return Res;
      }
    }

    // Level Zero has a much more explicit notion of command submission than
    // OpenCL. It doesn't happen unless the user submits a command list. We've
    // done it just above so the status is at least PI_EVENT_RUNNING.
    pi_int32 Result = PI_EVENT_RUNNING;

    // Make sure that we query a host-visible event only.
    // If one wasn't yet created then don't create it here as well, and
    // just conservatively return that event is not yet completed.
    std::shared_lock<pi_shared_mutex> EventLock(Event->Mutex);
    auto HostVisibleEvent = Event->HostVisibleEvent;
    if (Event->Completed) {
      Result = PI_EVENT_COMPLETE;
    } else if (HostVisibleEvent) {
      ze_result_t ZeResult;
      ZeResult =
          ZE_CALL_NOCHECK(zeEventQueryStatus, (HostVisibleEvent->ZeEvent));
      if (ZeResult == ZE_RESULT_SUCCESS) {
        Result = PI_EVENT_COMPLETE;
      }
    }
    return ReturnValue(pi_cast<pi_int32>(Result));
  }
```

Note `the status is at least PI_EVENT_RUNNING.` Unless this call is blocking until the task actually starts on the device (doesn't seem plausible), then again `event_command_status::running` actually just means 'passed on to the backend'.

Unfortunately, I think DPC++ isn't respecting the spec ([4.6.6.1](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_event_information_and_profiling_descriptors)) here:

> Indicates that the command has started running on the device but has not yet completed.

**Comment by abagusetty** [comment]
I meant to also check for the OpenCL backend soon as that can also provide some complete picture.

**Comment by abagusetty** [comment]
Update: The L0 backend seems to provide appropriate behavior on Intel GPUs(PVC). However, CUDA and HIP plugins still has the issue.
